src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,205 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal static class PercentEncodingHelpers
+    {
+        private static readonly char[] HexChars = "0123456789ABCDEF".ToCharArray();
+        private const int MaxUnicodeCharsReallocate = 40; // Maximum batch size when working with unicode characters
+        private const int MaxUtf8BytesPerUnicodeChar = 4;
+        private const int AsciiMaxValue = 127;
+
+        private static readonly byte[] PercentEncodingUnreservedBytes =
+        {
+            0x00, 0x00, 0x00, 0x00, 0xdf, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xff, 0xff, 0xff, 0x7f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+        };
+
+        public static string PercentEncode(string value)
```

Comment:
FYI, this is what Java does:
https://github.com/grpc/grpc-java/blob/b363f80764bcc8452ef54511b6c6d6b596f5f177/api/src/main/java/io/grpc/Status.java#L574

Reply:
Looks like a lot of allocations 😋 Not that matters since successful messages tend to not return a message with OK.